File: python/src/storage/adapters/mongodb/transaction.py
"""
MongoDB transaction implementation.

This module provides transaction management for MongoDB using Motor
with support for multi-document transactions and session management.
"""

from __future__ import annotations

import asyncio
from typing import Any, Dict, List, Optional

from motor.motor_asyncio import AsyncIOMotorClient, AsyncIOMotorClientSession, AsyncIOMotorDatabase
from pymongo import ReadPreference
from pymongo.read_concern import ReadConcern
from pymongo.write_concern import WriteConcern

from storage.adapters.mongodb.result import MongoDBResult, MongoDBRow
from storage.errors import (
    TransactionError,
    new_transaction_error,
    new_transaction_timeout_error,
    new_query_error,
)
from storage.interfaces import Transaction
from storage.types import (
    Command,
    ExecuteResult,
    IsolationLevel,
    Operation,
    OperationResult,
    OperationType,
    Query,
    TxOptions,
)


class MongoDBTransaction(Transaction):
    """MongoDB transaction implementation using Motor sessions."""
    
    def __init__(
        self,
        session: AsyncIOMotorClientSession,
        database: AsyncIOMotorDatabase,
        options: Optional[TxOptions] = None
    ) -> None:
        """Initialize MongoDB transaction.
        
        Args:
            session: Motor client session
            database: MongoDB database
            options: Transaction options
        """
        self._session = session
        self._database = database
        self._options = options
        self._active = True
    
    @classmethod
    async def create(
        cls,
        client: AsyncIOMotorClient,
        database: AsyncIOMotorDatabase,
        options: Optional[TxOptions] = None
    ) -> MongoDBTransaction:
        """Create a new MongoDB transaction.
        
        Args:
            client: MongoDB client
            database: MongoDB database
            options: Transaction options
            
        Returns:
            New transaction instance
            
        Raises:
            TransactionError: If transaction creation fails
        """
        try:
            # Start session
            session = await client.start_session()
            
            # Configure transaction options
            txn_options = {}
            
            if options:
                # Map isolation levels to MongoDB read concerns
                if options.isolation == IsolationLevel.READ_COMMITTED:
                    txn_options["read_concern"] = ReadConcern("majority")
                elif options.isolation == IsolationLevel.SERIALIZABLE:
                    txn_options["read_concern"] = ReadConcern("linearizable")
                
                # Set read preference for read-only transactions
                if options.read_only:
                    txn_options["read_preference"] = ReadPreference.SECONDARY_PREFERRED
                
                # Set write concern
                txn_options["write_concern"] = WriteConcern(w="majority")
            
            # Start transaction
            await session.start_transaction(**txn_options)
            
            return cls(session, database, options)
            
        except Exception as e:
            if 'session' in locals():
                await session.end_session()
            raise new_transaction_error(f"Failed to create MongoDB transaction: {e}") from e
    
    async def query(self, query: Query) -> MongoDBResult:
        """Execute a query within the transaction.
        
        Args:
            query: Query to execute
            
        Returns:
            Query results
            
        Raises:
            TransactionError: If transaction is not active
            QueryError: If query execution fails
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        try:
            # Extract collection name and build filter
            collection_name = self._extract_collection_name(query.sql)
            if not collection_name:
                raise new_query_error("Collection name not found in query")
            
            collection = self._database[collection_name]
            filter_doc = self._build_filter(query.parameters)
            
            # Execute find operation within session
            cursor = collection.find(filter_doc, session=self._session)
            
            return MongoDBResult(cursor)
            
        except Exception as e:
            if isinstance(e, (TransactionError, new_query_error().__class__)):
                raise
            raise new_query_error(f"MongoDB transaction query failed: {e}") from e
    
    async def query_one(self, query: Query) -> MongoDBRow:
        """Execute a query and return a single row within the transaction.
        
        Args:
            query: Query to execute
            
        Returns:
            Single query result
            
        Raises:
            TransactionError: If transaction is not active
            QueryError: If query execution fails
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        try:
            # Extract collection name and build filter
            collection_name = self._extract_collection_name(query.sql)
            if not collection_name:
                raise new_query_error("Collection name not found in query")
            
            collection = self._database[collection_name]
            filter_doc = self._build_filter(query.parameters)
            
            # Execute findOne operation within session
            document = await collection.find_one(filter_doc, session=self._session)
            if document is None:
                raise new_query_error("No document found")
            
            return MongoDBRow(document)
            
        except Exception as e:
            if isinstance(e, (TransactionError, new_query_error().__class__)):
                raise
            raise new_query_error(f"MongoDB transaction query_one failed: {e}") from e
    
    async def execute(self, command: Command) -> ExecuteResult:
        """Execute a command within the transaction.
        
        Args:
            command: Command to execute
            
        Returns:
            Execution result
            
        Raises:
            TransactionError: If transaction is not active
            QueryError: If command execution fails
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        try:
            # Route to appropriate execution method based on command type
            if command.type.value == "insert":
                return await self._execute_insert(command)
            elif command.type.value == "update":
                return await self._execute_update(command)
            elif command.type.value == "delete":
                return await self._execute_delete(command)
            else:
                raise new_query_error(f"Unsupported command type: {command.type}")
                
        except Exception as e:
            if isinstance(e, (TransactionError, new_query_error().__class__)):
                raise
            raise new_query_error(f"MongoDB transaction command failed: {e}") from e
    
    async def batch(self, operations: List[Operation]) -> List[OperationResult]:
        """Execute multiple operations within the transaction.
        
        Args:
            operations: List of operations to execute
            
        Returns:
            List of operation results
            
        Raises:
            TransactionError: If transaction is not active
            QueryError: If batch execution fails
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        results = []
        
        # Group operations by collection for bulk operations
        collection_ops: Dict[str, List[tuple[int, Operation]]] = {}
        
        for i, op in enumerate(operations):
            collection_name = self._extract_collection_name(op.query.sql)
            if not collection_name:
                results.append(OperationResult(
                    success=False,
                    error=new_query_error("Collection name not found in operation")
                ))
                continue
            
            if collection_name not in collection_ops:
                collection_ops[collection_name] = []
            collection_ops[collection_name].append((i, op))
        
        # Initialize results list
        results = [OperationResult(success=False) for _ in operations]
        
        # Execute bulk operations for each collection within session
        for collection_name, ops in collection_ops.items():
            try:
                collection = self._database[collection_name]
                bulk_ops = []
                
                for _, op in ops:
                    if op.type == OperationType.INSERT:
                        doc = self._build_document(op.query.parameters)
                        bulk_ops.append({"insertOne": {"document": doc}})
                    elif op.type == OperationType.UPDATE:
                        mid = len(op.query.parameters) // 2
                        filter_doc = self._build_filter(op.query.parameters[:mid])
                        update_doc = {"$set": self._build_document(op.query.parameters[mid:])}
                        bulk_ops.append({"updateOne": {"filter": filter_doc, "update": update_doc}})
                    elif op.type == OperationType.DELETE:
                        filter_doc = self._build_filter(op.query.parameters)
                        bulk_ops.append({"deleteOne": {"filter": filter_doc}})
                
                if bulk_ops:
                    await collection.bulk_write(bulk_ops, session=self._session)
                    
                    # Mark all operations for this collection as successful
                    for idx, _ in ops:
                        results[idx] = OperationResult(success=True)
                        
            except Exception as e:
                # Mark all operations for this collection as failed
                for idx, _ in ops:
                    results[idx] = OperationResult(
                        success=False,
                        error=new_query_error(f"Transaction batch operation failed: {e}")
                    )
        
        return results
    
    async def commit(self) -> None:
        """Commit the transaction.
        
        Raises:
            TransactionError: If commit fails or transaction is not active
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        try:
            await self._session.commit_transaction()
            self._active = False
            await self._session.end_session()
        except Exception as e:
            self._active = False
            await self._session.end_session()
            raise new_transaction_error(f"Failed to commit MongoDB transaction: {e}") from e
    
    async def rollback(self) -> None:
        """Rollback the transaction.
        
        Raises:
            TransactionError: If rollback fails or transaction is not active
        """
        if not self._active:
            raise new_transaction_error("Transaction is not active")
        
        try:
            await self._session.abort_transaction()
            self._active = False
            await self._session.end_session()
        except Exception as e:
            self._active = False
            await self._session.end_session()
            raise new_transaction_error(f"Failed to rollback MongoDB transaction: {e}") from e
    
    # Helper methods
    
    def _extract_collection_name(self, sql: str) -> str:
        """Extract collection name from query SQL.
        
        This is a simplified implementation - in practice, you'd have a proper parser.
        """
        if not sql:
            return ""
        
        # For now, assume the collection name is passed in a specific format
        # In a real implementation, you'd parse SQL-like syntax or use MongoDB-specific format
        return "default_collection"
    
    def _build_filter(self, params: List[Any]) -> Dict[str, Any]:
        """Build MongoDB filter document from parameters."""
        filter_doc = {}
        
        # Simple parameter mapping - in practice, you'd have sophisticated query building
        for i in range(0, len(params), 2):
            if i + 1 < len(params):
                key = params[i]
                value = params[i + 1]
                if isinstance(key, str):
                    filter_doc[key] = value
        
        return filter_doc
    
    def _build_document(self, params: List[Any]) -> Dict[str, Any]:
        """Build MongoDB document from parameters."""
        return self._build_filter(params)
    
    async def _execute_insert(self, command: Command) -> ExecuteResult:
        """Execute an insert command within transaction."""
        collection_name = self._extract_collection_name(command.sql)
        if not collection_name:
            raise new_query_error("Collection name not found in command")
        
        collection = self._database[collection_name]
        document = self._build_document(command.parameters)
        
        result = await collection.insert_one(document, session=self._session)
        
        return ExecuteResult(
            rows_affected=1,
            last_insert_id=str(result.inserted_id) if result.inserted_id else None
        )
    
    async def _execute_update(self, command: Command) -> ExecuteResult:
        """Execute an update command within transaction."""
        collection_name = self._extract_collection_name(command.sql)
        if not collection_name:
            raise new_query_error("Collection name not found in command")
        
        collection = self._database[collection_name]
        
        # Split parameters into filter and update parts
        mid = len(command.parameters) // 2
        filter_doc = self._build_filter(command.parameters[:mid])
        update_doc = {"$set": self._build_document(command.parameters[mid:])}
        
        result = await collection.update_many(filter_doc, update_doc, session=self._session)
        
        return ExecuteResult(
            rows_affected=result.modified_count,
            last_insert_id=None
        )
    
    async def _execute_delete(self, command: Command) -> ExecuteResult:
        """Execute a delete command within transaction."""
        collection_name = self._extract_collection_name(command.sql)
        if not collection_name:
            raise new_query_error("Collection name not found in command")
        
        collection = self._database[collection_name]
        filter_doc = self._build_filter(command.parameters)
        
        result = await collection.delete_many(filter_doc, session=self._session)
        
        return ExecuteResult(
            rows_affected=result.deleted_count,
            last_insert_id=None
        )

File: internal/infrastructure/adapters/mongodb/transaction.go
package mongodb

import (
	"context"
	"fmt"
	"sync"

	"github.com/AnandSGit/HybridCache.io/internal/domain"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
)

// MongoDBTransaction implements the domain.Transaction interface for MongoDB
type MongoDBTransaction struct {
	session  mongo.Session
	database *mongo.Database
	adapter  *Adapter
	active   bool
	mu       sync.RWMutex
}

// Query executes a query within the transaction
func (t *MongoDBTransaction) Query(ctx context.Context, query domain.Query) (domain.Result, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	if !t.active {
		return nil, domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	// Execute query within session context
	sessionCtx := mongo.NewSessionContext(ctx, t.session)

	// Extract collection name from query
	collectionName := t.extractCollectionName(query.SQL)
	if collectionName == "" {
		return nil, domain.NewQueryError("INVALID_QUERY", "collection name not found in query")
	}

	collection := t.database.Collection(collectionName)

	// Build filter from parameters
	filter := t.buildFilter(query.Parameters)

	// Execute find operation within transaction
	cursor, err := collection.Find(sessionCtx, filter)
	if err != nil {
		return nil, domain.NewQueryError("QUERY_EXECUTION_FAILED", err.Error()).WithCause(err)
	}

	return &MongoDBResult{cursor: cursor}, nil
}

// QueryOne executes a query and returns a single row within the transaction
func (t *MongoDBTransaction) QueryOne(ctx context.Context, query domain.Query) (domain.Row, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	if !t.active {
		return nil, domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	// Execute query within session context
	sessionCtx := mongo.NewSessionContext(ctx, t.session)

	// Extract collection name from query
	collectionName := t.extractCollectionName(query.SQL)
	if collectionName == "" {
		return nil, domain.NewQueryError("INVALID_QUERY", "collection name not found in query")
	}

	collection := t.database.Collection(collectionName)

	// Build filter from parameters
	filter := t.buildFilter(query.Parameters)

	// Execute findOne operation within transaction
	var result bson.M
	err := collection.FindOne(sessionCtx, filter).Decode(&result)
	if err != nil {
		if err == mongo.ErrNoDocuments {
			return nil, domain.NewDataError("NO_ROWS", "no document found")
		}
		return nil, domain.NewQueryError("QUERY_EXECUTION_FAILED", err.Error()).WithCause(err)
	}

	return &MongoDBRow{document: result}, nil
}

// Execute executes a command within the transaction
func (t *MongoDBTransaction) Execute(ctx context.Context, command domain.Command) (domain.ExecuteResult, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	if !t.active {
		return domain.ExecuteResult{}, domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	// Execute command within session context
	sessionCtx := mongo.NewSessionContext(ctx, t.session)

	// Parse command type and execute appropriate MongoDB operation
	switch command.Type {
	case domain.CommandTypeInsert:
		return t.executeInsert(sessionCtx, command)
	case domain.CommandTypeUpdate:
		return t.executeUpdate(sessionCtx, command)
	case domain.CommandTypeDelete:
		return t.executeDelete(sessionCtx, command)
	default:
		return domain.ExecuteResult{}, domain.NewQueryError("UNSUPPORTED_COMMAND", "command type not supported in transaction")
	}
}

// Batch executes multiple operations within the transaction
func (t *MongoDBTransaction) Batch(ctx context.Context, operations []domain.Operation) ([]domain.OperationResult, error) {
	t.mu.RLock()
	defer t.mu.RUnlock()

	if !t.active {
		return nil, domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	// Execute batch within session context
	sessionCtx := mongo.NewSessionContext(ctx, t.session)
	results := make([]domain.OperationResult, len(operations))

	// Group operations by collection for bulk operations
	collectionOps := make(map[string][]domain.Operation)
	opIndexes := make(map[string][]int)

	for i, op := range operations {
		var collectionName string
		if op.Type == domain.OperationTypeQuery {
			collectionName = t.extractCollectionName(op.Query.SQL)
		} else if op.Type == domain.OperationTypeCommand {
			collectionName = t.extractCollectionName(op.Command.SQL)
		}

		if collectionName == "" {
			results[i] = domain.OperationResult{
				Index: i,
				Error: domain.NewQueryError("INVALID_OPERATION", "collection name not found"),
			}
			continue
		}
		collectionOps[collectionName] = append(collectionOps[collectionName], op)
		opIndexes[collectionName] = append(opIndexes[collectionName], i)
	}

	// Execute bulk operations for each collection
	for collectionName, ops := range collectionOps {
		collection := t.database.Collection(collectionName)
		indexes := opIndexes[collectionName]

		// Build bulk write operations
		var bulkOps []mongo.WriteModel
		for _, op := range ops {
			if op.Type == domain.OperationTypeCommand {
				switch op.Command.Type {
				case domain.CommandTypeInsert:
					doc := t.buildDocument(op.Command.Parameters)
					bulkOps = append(bulkOps, mongo.NewInsertOneModel().SetDocument(doc))
				case domain.CommandTypeUpdate:
					filter := t.buildFilter(op.Command.Parameters[:len(op.Command.Parameters)/2])
					update := t.buildUpdate(op.Command.Parameters[len(op.Command.Parameters)/2:])
					bulkOps = append(bulkOps, mongo.NewUpdateOneModel().SetFilter(filter).SetUpdate(update))
				case domain.CommandTypeDelete:
					filter := t.buildFilter(op.Command.Parameters)
					bulkOps = append(bulkOps, mongo.NewDeleteOneModel().SetFilter(filter))
				}
			}
		}

		// Execute bulk write within transaction
		if len(bulkOps) > 0 {
			_, err := collection.BulkWrite(sessionCtx, bulkOps)
			if err != nil {
				// Mark all operations for this collection as failed
				for _, idx := range indexes {
					results[idx] = domain.OperationResult{
						Index: idx,
						Error: domain.NewQueryError("BULK_WRITE_FAILED", err.Error()).WithCause(err),
					}
				}
			} else {
				// Mark all operations for this collection as successful
				for _, idx := range indexes {
					results[idx] = domain.OperationResult{
						Index:  idx,
						Result: domain.ExecuteResult{RowsAffected: 1},
					}
				}
			}
		}
	}

	return results, nil
}

// Commit commits the transaction
func (t *MongoDBTransaction) Commit() error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if !t.active {
		return domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	ctx := context.Background()
	err := t.session.CommitTransaction(ctx)
	if err != nil {
		return domain.NewTransactionError("COMMIT_FAILED", "failed to commit MongoDB transaction").WithCause(err)
	}

	t.active = false
	t.session.EndSession(ctx)
	return nil
}

// Rollback rolls back the transaction
func (t *MongoDBTransaction) Rollback() error {
	t.mu.Lock()
	defer t.mu.Unlock()

	if !t.active {
		return domain.NewTransactionError("TRANSACTION_CLOSED", "transaction is not active")
	}

	ctx := context.Background()
	err := t.session.AbortTransaction(ctx)
	if err != nil {
		return domain.NewTransactionError("ROLLBACK_FAILED", "failed to rollback MongoDB transaction").WithCause(err)
	}

	t.active = false
	t.session.EndSession(ctx)
	return nil
}

// Savepoint creates a savepoint (not supported in MongoDB)
func (t *MongoDBTransaction) Savepoint(name string) error {
	return domain.NewTransactionError("SAVEPOINT_NOT_SUPPORTED", "MongoDB does not support savepoints")
}

// RollbackToSavepoint rolls back to a savepoint (not supported in MongoDB)
func (t *MongoDBTransaction) RollbackToSavepoint(name string) error {
	return domain.NewTransactionError("SAVEPOINT_NOT_SUPPORTED", "MongoDB does not support savepoints")
}

// IsActive returns whether the transaction is active
func (t *MongoDBTransaction) IsActive() bool {
	t.mu.RLock()
	defer t.mu.RUnlock()
	return t.active
}

// ID returns the transaction ID
func (t *MongoDBTransaction) ID() string {
	// MongoDB sessions don't have a simple string ID
	// Return a placeholder for now
	return "mongodb-transaction"
}

// Helper methods (reuse from storage.go)

// extractCollectionName extracts collection name from query SQL
func (t *MongoDBTransaction) extractCollectionName(sql string) string {
	// Simple extraction - in practice, you'd have a proper parser
	if sql == "" {
		return ""
	}
	return "default_collection"
}

// buildFilter builds a MongoDB filter from parameters
func (t *MongoDBTransaction) buildFilter(params []interface{}) bson.M {
	filter := bson.M{}

	for i := 0; i < len(params); i += 2 {
		if i+1 < len(params) {
			if key, ok := params[i].(string); ok {
				filter[key] = params[i+1]
			}
		}
	}

	return filter
}

// buildDocument builds a MongoDB document from parameters
func (t *MongoDBTransaction) buildDocument(params []interface{}) bson.M {
	return t.buildFilter(params)
}

// buildUpdate builds a MongoDB update document from parameters
func (t *MongoDBTransaction) buildUpdate(params []interface{}) bson.M {
	update := bson.M{"$set": t.buildFilter(params)}
	return update
}

// executeInsert executes an insert command within transaction
func (t *MongoDBTransaction) executeInsert(ctx context.Context, command domain.Command) (domain.ExecuteResult, error) {
	collectionName := t.extractCollectionName(command.SQL)
	if collectionName == "" {
		return domain.ExecuteResult{}, domain.NewQueryError("INVALID_COMMAND", "collection name not found")
	}

	collection := t.database.Collection(collectionName)
	document := t.buildDocument(command.Parameters)

	result, err := collection.InsertOne(ctx, document)
	if err != nil {
		return domain.ExecuteResult{}, domain.NewQueryError("INSERT_FAILED", err.Error()).WithCause(err)
	}

	// Convert InsertedID to int64 (MongoDB ObjectID as hash)
	var lastInsertID int64
	if result.InsertedID != nil {
		// For MongoDB, we'll use a hash of the ObjectID string representation
		lastInsertID = int64(fmt.Sprintf("%v", result.InsertedID)[0:8][0]) // Simplified conversion
	}

	return domain.ExecuteResult{
		RowsAffected: 1,
		LastInsertID: lastInsertID,
	}, nil
}

// executeUpdate executes an update command within transaction
func (t *MongoDBTransaction) executeUpdate(ctx context.Context, command domain.Command) (domain.ExecuteResult, error) {
	collectionName := t.extractCollectionName(command.SQL)
	if collectionName == "" {
		return domain.ExecuteResult{}, domain.NewQueryError("INVALID_COMMAND", "collection name not found")
	}

	collection := t.database.Collection(collectionName)

	mid := len(command.Parameters) / 2
	filter := t.buildFilter(command.Parameters[:mid])
	update := t.buildUpdate(command.Parameters[mid:])

	result, err := collection.UpdateMany(ctx, filter, update)
	if err != nil {
		return domain.ExecuteResult{}, domain.NewQueryError("UPDATE_FAILED", err.Error()).WithCause(err)
	}

	return domain.ExecuteResult{
		RowsAffected: result.ModifiedCount,
	}, nil
}

// executeDelete executes a delete command within transaction
func (t *MongoDBTransaction) executeDelete(ctx context.Context, command domain.Command) (domain.ExecuteResult, error) {
	collectionName := t.extractCollectionName(command.SQL)
	if collectionName == "" {
		return domain.ExecuteResult{}, domain.NewQueryError("INVALID_COMMAND", "collection name not found")
	}

	collection := t.database.Collection(collectionName)
	filter := t.buildFilter(command.Parameters)

	result, err := collection.DeleteMany(ctx, filter)
	if err != nil {
		return domain.ExecuteResult{}, domain.NewQueryError("DELETE_FAILED", err.Error()).WithCause(err)
	}

	return domain.ExecuteResult{
		RowsAffected: result.DeletedCount,
	}, nil
}
